src/whir/batch_proof.rs

```diff
@@ -0,0 +1,423 @@
+    /// * `f_a` - Evaluations of the first polynomial
+    /// * `f_b` - Evaluations of the second polynomial
+    /// * `z_a` - Evaluation point for f_a
+    /// * `z_b` - Evaluation point for f_b
+    /// * `v_a` - Claimed evaluation f_a(z_a)
+    /// * `v_b` - Claimed evaluation f_b(z_b)
+    /// * `alpha` - Batching randomness
+    ///
+    /// # Returns
+    ///
+    /// A tuple containing:
+    /// * `SumcheckSingle<F, EF>` - The folded sumcheck state
+    /// * `EF` - The folding challenge r_0
+    #[allow(clippy::too_many_arguments)]
+    fn selector_round(
+        &self,
+        selector_data: &mut SumcheckData<EF, F>,
+        challenger: &mut Challenger,
+        f_a: &EvaluationsList<F>,
+        f_b: &EvaluationsList<F>,
+        z_a: &MultilinearPoint<EF>,
+        z_b: &MultilinearPoint<EF>,
+        v_a: EF,
+        v_b: EF,
+        alpha: EF,
+    ) -> (SumcheckSingle<F, EF>, EF) {
+        // Create combined polynomial and weights for selector round:
+        // The combined polynomial f_c(X, b) = X·f_a(b) + (1-X)·f_b(b) over {0,1}^{m+1}
+        // has evaluations [f_b | f_a] (first half is f_b at X=0, second half is f_a at X=1)
+        //
+        // The combined weight w(X, b) = X·eq(b, z_a) + α(1-X)·eq(b, z_b)
+        // has evaluations [α·eq(·, z_b) | eq(·, z_a)]
+
+        // Build combined polynomial: [f_b | f_a]
+        let combined_evals: Vec<F> = f_b
+            .as_slice()
+            .iter()
+            .chain(f_a.as_slice().iter())
+            .copied()
+            .collect();
+        let combined_poly = EvaluationsList::new(combined_evals);
+
+        // Build combined weights: [α·eq(·, z_b) | eq(·, z_a)]
+        let eq_z_b = EvaluationsList::new_from_point(z_b.as_slice(), alpha);
+        let eq_z_a = EvaluationsList::new_from_point(z_a.as_slice(), EF::ONE);
+        let combined_weights_vec: Vec<EF> = eq_z_b
+            .as_slice()
+            .iter()
+            .chain(eq_z_a.as_slice().iter())
+            .copied()
+            .collect();
+        let combined_weights = EvaluationsList::new(combined_weights_vec);
+
+        // Compute sumcheck polynomial coefficients:
+        // h(X) = Σ_b f_c(X, b) · w(X, b)
+        // We compute c0 = h(0) and c2 (quadratic coefficient)
+        let (c0, c2) = combined_poly.sumcheck_coefficients(&combined_weights);
+
+        // Sanity check: c0 should equal α·v_b
+        debug_assert_eq!(c0, alpha * v_b, "c0 = h(0) should equal α·v_b");
+
+        // Observe Fiat-Shamir
+        let pow_bits = self.starting_folding_pow_bits;
+        let r_0 = selector_data.observe_and_sample::<Challenger, F>(challenger, c0, c2, pow_bits);
+
+        // Fold the polynomial and weights:
+        // g(b) = r_0·f_a(b) + (1-r_0)·f_b(b)
+        // w'(b) = r_0·eq(b, z_a) + α(1-r_0)·eq(b, z_b)
+        // σ' = h(r_0)
+
+        // Folded polynomial in extension field
+        let g = EvaluationsList::linear_combination(f_a, r_0, f_b, EF::ONE - r_0);
```

Comment:
Similarly maybe it is possible to combine f_a and f_b onto f_a by getting f_a with mut value